controllers/openstackservergroup_controller.go

```diff
@@ -0,0 +1,214 @@
+
+	log.Info("OpenStackServerGroup is about to reconcile")
+
+	if annotations.HasPaused(openStackServerGroup) {
+		log.Info("OpenStackServerGroup is marked as paused. Won't reconcile")
+		return ctrl.Result{}, nil
+	}
+
+	// Initialize the patch helper
+	patchHelper, err := patch.NewHelper(openStackServerGroup, r.Client)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Always patch the openStackServerGroup when exiting this function so we can persist any changes.
+	defer func() {
+		if err := patchServerGroup(ctx, patchHelper, openStackServerGroup); err != nil {
+			result = ctrl.Result{}
+			reterr = kerrors.NewAggregate([]error{reterr, err})
+		}
+	}()
+
+	scope, err := r.ScopeFactory.NewClientScopeFromServerGroup(ctx, r.Client, openStackServerGroup, r.CaCertificates, log)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Handle deleted servergroups
+	if !openStackServerGroup.DeletionTimestamp.IsZero() {
+		return r.reconcileDelete(scope, openStackServerGroup)
+	}
+
+	// Handle non-deleted servergroups
+	return r.reconcileNormal(ctx, scope, openStackServerGroup)
+}
+
+func patchServerGroup(ctx context.Context, patchHelper *patch.Helper, openStackServerGroup *infrav1.OpenStackServerGroup, options ...patch.Option) error {
+	return patchHelper.Patch(ctx, openStackServerGroup, options...)
+}
+
+func (r *OpenStackServerGroupReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(
+			&infrav1.OpenStackServerGroup{},
+			builder.WithPredicates(
+				predicate.Funcs{
+					// Avoid reconciling if the event triggering the reconciliation is related to incremental status updates
+					UpdateFunc: func(e event.UpdateEvent) bool {
+						oldServerGroup := e.ObjectOld.(*infrav1.OpenStackServerGroup).DeepCopy()
+						newServerGroup := e.ObjectNew.(*infrav1.OpenStackServerGroup).DeepCopy()
+						oldServerGroup.Status = infrav1.OpenStackServerGroupStatus{}
+						newServerGroup.Status = infrav1.OpenStackServerGroupStatus{}
+						oldServerGroup.ObjectMeta.ResourceVersion = ""
+						newServerGroup.ObjectMeta.ResourceVersion = ""
+						return !reflect.DeepEqual(oldServerGroup, newServerGroup)
+					},
+				},
+			),
+		).
+		WithEventFilter(predicates.ResourceNotPausedAndHasFilterLabel(ctrl.LoggerFrom(ctx), r.WatchFilterValue)).
+		Complete(r)
+}
+
+func (r *OpenStackServerGroupReconciler) reconcileDelete(scope scope.Scope, openStackServerGroup *infrav1.OpenStackServerGroup) (result ctrl.Result, reterr error) {
+	scope.Logger().Info("Reconciling ServerGroup delete")
+
+	serverGroupName := openStackServerGroup.Name
+
+	computeService, err := compute.NewService(scope)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	// Get the servergroup by name, even if our K8s resource already has the ID field set.
+	// TODO(dalees): If this returns a 404 do we try to delete with existing UUID? Do we just assume success?
```

Comment:
I think we should look up by ID and only then fallback to looking up by name. IDs are safe in case of duplicate names.

Reply:
Ok, happy to change this - I can see it will lead to less problems if a duplicate named resource was created after this managed one.

Reply:
The delete operation no longer looks up by name, and now only deletes server groups from OpenStack if there is a stored ID.